### PR TITLE
added more methods / properties, world account ...

### PR DESCRIPTION
**Pull request by Maxthegreat99**
... Added extentions for TSplayer related to bank accounts, balance now can only be set via methods that force transactions, started the work on a system currency ...

# To do
- ~~Fix the commands to not refer to `player.Account` as it may cause a null reference with manipulating server (to verify)~~
- ~~Add localization support for the plugin's strings~~
- ~~Port redundant / duplicate methods to their proper classes (ex: `Helper` methods -> `Api`, `tranferTo()` -> `BankAccount`)~~
- (Potentially) implement IDs for currencies
- Create async counterparts of methods and implement them where needed
- ~~implement Accounts reset after a new world has been detected, most likely ask the user for input for confirmation to the action even if the setting is set to true.~~
- id currencies
- log transactions via a queue that only logs after making sure there are no incoming transactions
- ~~private the DB and cache bank accounts.~~
- ~~instead of a bank acc for each currency make bank accs have a list for currencies~~
- ~~clean logs depending on how many log files are in the logs folder.~~

### CODE REVIEW

File: TSEconomy/TSEconomy.cs [hunk 1]
```diff
@@ -21,21 +22,28 @@ public TSEconomy(Main game) : base(game)
...
+
+        public void OnInitialize(EventArgs args)
         {
             // load our config file
             Configuration.Configuration.Load();
 
             // init db
             DB.InitializeDB(Config.UseMySQL);
 
-            // register commands
```
Comment (RenderBr) [inline TSEconomy/TSEconomy.cs]: is this change necessary? i may be wrong but GameInitialize occurs when the world loads, i dont see the point of this tbh

File: TSEconomy/Database/Models/TransactionProperties.cs [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace TSEconomy.Database.Models
+{
+    public enum TransactionProperties
+    {
+        set = 1,
+        add = 1 << 1
```
Comment (RenderBr) [inline TSEconomy/Database/Models/TransactionProperties.cs]: enum members should follow PascalCase

File: TSEconomy/Database/Models/Transaction.cs
Comment (RenderBr) [inline TSEconomy/Database/Models/Transaction.cs:1]: this new column cannot be serialized into our database yet

File: TSEconomy/Logging/TransactionLogging.cs [hunk 3]
```diff
@@ -35,7 +35,9 @@ public static void Log(Transaction trans)
 
             using (StreamWriter writer = File.AppendText(sessionLogFile))
             {
-                writer.WriteLine($"[{trans.Timestamp}] : {trans.TransactionDetails}");
+                string flag = (trans.Flags & TransactionProperties.set) == TransactionProperties.set ? "Set" : "Add";
```
Comment (RenderBr) [inline TSEconomy/Logging/TransactionLogging.cs]: what is this doing?

File: TSEconomy/TSEconomy.cs (hunk 1 above)
Comment (Maxthegreat99) [reply TSEconomy/TSEconomy.cs]: planning on making the plugin show up startup messages which usually should be sent after plugins are done initializing to avoid tshock's default startup message, i think when the world loads is GamePostInitialize, i have to test.

File: TSEconomy/Logging/TransactionLogging.cs (hunk 3 above)
Comment (Maxthegreat99) [reply TSEconomy/Logging/TransactionLogging.cs]: specifying whether or not the transaction is a Set type or an Add type in the log message, since the plugin is gonna get used as an API, we cant be sure whether or not other plugins will specify this in their details

File: TSEconomy/Database/Models/TransactionProperties.cs (hunk 2 above)
Comment (Maxthegreat99) [reply TSEconomy/Database/Models/TransactionProperties.cs]: sorry will change this next commit, was quite braindead while writing those

File: TSEconomy/Database/Models/Transaction.cs
Comment (Maxthegreat99) [reply TSEconomy/Database/Models/Transaction.cs:1]: speaking of which, should we put transactions into the database? we wont be checking whether or not the transactions add up since its required to make a transaction to set / add to the balance, we can warn players if anything suspicious is happening within the transaction while it gets logged

File: TSEconomy/Helpers.cs
Comment (Maxthegreat99) [inline TSEconomy/Helpers.cs:1]: is there a difference between `Helper` & `Api`?

File: TSEconomy/TSEconomy.cs (hunk 1 above)
Comment (RenderBr) [reply TSEconomy/TSEconomy.cs]: you may be right -- not totally sure. idk, ideally i'd like for them to keep as little bloat as possible in the codebase, and it doesn't make a ton of sense to split up initialization methods into hooks when it doesn't change functionality for the better. unless i am glossing over some benefit

File: TSEconomy/Database/Models/TransactionProperties.cs (hunk 2 above)
Comment (RenderBr) [reply TSEconomy/Database/Models/TransactionProperties.cs]: ur good dw dw

File: TSEconomy/Database/Models/Transaction.cs
Comment (RenderBr) [reply TSEconomy/Database/Models/Transaction.cs:1]: i guess we do not especially need to, maybe we can make this config based in case someone plans on making an external service that pulls transaction history... this is the only usecase i can see where it would be useful. it will save us some perf if we allow it to be disabled when not necessary

File: TSEconomy/Logging/TransactionLogging.cs (hunk 3 above)
Comment (RenderBr) [reply TSEconomy/Logging/TransactionLogging.cs]: > specifying whether or not the transaction is a Set type or an Add type in the log message, since the plugin is gonna get used as an API, we cant be sure whether or not other plugins will specify this in their details

ah okay, i see thanks

File: TSEconomy/Helpers.cs
Comment (RenderBr) [reply TSEconomy/Helpers.cs:1]: > is there a difference between `Helper` & `Api`?

i mean the key difference here is the API is a class used for directly interacting with our plugin, whereas Helpers is essentially just a utilities class for reducing boilerplate. i am up to suggestions on how to make things simpler

**Review by RenderBr — APPROVED** [review]

File: TSEconomy/TSEconomy.cs (hunk 1 above)
Comment (Maxthegreat99) [reply TSEconomy/TSEconomy.cs]: you can check AntiVPN: https://github.com/Maxthegreat99/AntiVPN-TShock/blob/master/Main.cs it sends startup messages after all plugins initialized and is hooking GameInitialize. As for splitting intialization when i wrote this i mostly thought since onInitialize was later on gonna contain some initialization code later on anyway (stuff like checking if world ids are the same etc..) thought it would be good to load it there. now that i think of it since we are making an api it would probably be better loading it when the plugin is actually loading and making sure the plugin loads first as to give a chance for other plugins to do their work when initializing, gonna change it right now

File: TSEconomy/Database/Models/Transaction.cs
Comment (Maxthegreat99) [reply TSEconomy/Database/Models/Transaction.cs:1]: alr then, ill plan to add a squashJournal functionality

File: TSEconomy/Helpers.cs
Comment (Maxthegreat99) [reply TSEconomy/Helpers.cs:1]: alr, will move some of my methods to Helper if needed
